core/src/main/java/com/linecorp/armeria/client/DefaultEventLoopScheduler.java

```diff
@@ -0,0 +1,443 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static it.unimi.dsi.fastutil.Hash.FAST_LOAD_FACTOR;
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.ReleasableHolder;
+
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+
+final class DefaultEventLoopScheduler implements EventLoopScheduler {
+
+    private static final long CLEANUP_INTERVAL_NANOS = Duration.ofMinutes(1).toNanos();
+
+    private final List<EventLoop> eventLoops;
+
+    private final int maxNumEventLoops;
+
+    private final Object2IntMap<String> maxNumEventLoopsMap;
+
+    private final Map<AuthorityWithTLS, State> states = new ConcurrentHashMap<>();
+
+    private final AtomicInteger currentIndex = new AtomicInteger();
+
+    private int counter;
+    private volatile long lastCleanupTimeNanos = System.nanoTime();
+
+    DefaultEventLoopScheduler(EventLoopGroup eventLoopGroup, int maxNumEventLoops,
+                              Map<String, Integer> maxNumEventLoopsMap) {
+        eventLoops = Streams.stream(eventLoopGroup)
+                            .map(EventLoop.class::cast)
+                            .collect(toImmutableList());
+        checkArgument(maxNumEventLoops <= eventLoops.size(),
+                      "maxNumEventLoopsPerEndpoint: %s (expected: <= %s)", maxNumEventLoops,
+                      eventLoops.size());
+        if (maxNumEventLoops <= 0) {
+            this.maxNumEventLoops = eventLoops.size();
```

Comment:
What would be the sensible default? `1`? /cc @anuraaga

Reply:
Had a chat with @trustin, and we concluded `1` for `h2`, `h2c`, `HTTP` and `HTTPS` and `eventLoops.size()` for `h1` and `h1c`. :-)

Reply:
Thanks that sounds even better :)

Reply:
Had a chat with @trustin again, and we decide to use `1` for all session protocols. A user would get confused when he/she sees different behavior when sending requests to "http://a.com" and "h1c://b.com". Please forgive my impatience. 🙇